### ISSUE TITLE
How to use pykaldi to decode words from pytorch phoneme model

### ISSUE THREAD

**Issue by IanMGriff** [pykaldi/pykaldi#293]
I have a trained acoustic model (Pytorch RNN) that outputs frame wise phone likelihoods, and a fst composed of a token, lexicon, and grammar fst. I have tried to use pykaldi to connect my RNN and FST for an end-to-end framework, but my FST seems to output the same result no matter what input it is given.  I am not able to share the full code, but this is an example of all relevant parts in my decoding approach 

```
from kaldi.asr import FasterRecognizer
from kaldi.decoder import FasterDecoderOptions, FasterDecoder
from kaldi.matrix import Matrix
import kaldi.fstext as fst 

from kaldi.fstext import utils as _fst_utils
import kaldi.decoder as dec


fst_path = 'path/to/TLG.fst'
symbols_path = 'path/to/words.txt'

decoder_opts = FasterDecoderOptions()
decoder_opts.beam = 10
decoder_opts.max_active = 100

fst_ = fst.read_fst_kaldi(fst_path)
kaldi_decoder = FasterDecoder(fst_, decoder_opts)
symbols = fst.SymbolTable.read_text(symbols_path)


def get_words_from_phones(out_for_pykaldi, kaldi_fst_decoder, symbols=symbols):
    loglikes = Matrix(out_for_pykaldi)
    decodable = dec.DecodableMatrixScaled(loglikes,0) # No acoustic model scaling 
    kaldi_fst_decoder.decode(decodable)
    best_path = kaldi_fst_decoder.get_best_path()
    ali, words, weight = _fst_utils.get_linear_symbol_sequence(best_path)
    return " ".join(fst.indices_to_symbols(symbols, words))

# I have verified the torch model works fine 

import torch_model
import asr_data 

ckpt_path = 'path/to/torch_ckpt.  

trained_pytorch_model = torch_model(ckpt = ckpt_path)

data = asr_data()

in, label = data.next() 

output = trained_pytorch_model(in) 

output_for_pykaldi = output.cpu().detatch().numpy().squeeze() # shape is now [frames, phoneme classes]

words = get_words_from_phones(output_for_pykaldi, kaldi_decoder, symbols) 

```

This always returns
> 'HE SAID'

No matter what input I pass into the FST.  

Questions:
1) Is this even the right way to use an FST with pykaldi? 
2) how can I debug the kaldi decoder to find out why the FST returns the same result.

I am fairly new to FSTs and kaldi, so I apologize if I am missing something obvious.  

Thank you for the help!

**Comment by IanMGriff** [comment]
Wanted to follow up with a solution to close this issue.

Make sure to set scaling for 1 to get no scaling behavior. 

```
def get_words_from_phones(out_for_pykaldi, kaldi_fst_decoder, symbols=symbols):
    loglikes = Matrix(out_for_pykaldi)
    decodable = dec.DecodableMatrixScaled(loglikes, 1) # Set to 1 for no model scaling
    ...
```
***
Also, some notes on connecting PyTorch models with Kaldi FSTs like this:
 - Make sure you are actually sending log-likelihoods to the TLG.fst for decoding.  I.e. if you are using a DeepSpeech2 model,  the final output will be likelihoods from a softmax layer (in evaluation mode). You'll need to convert those likelihood outputs to log likelihoods.
 -  Make sure the outputs getting set to your FSTs are of shape `[frames, phone_classes]`. If you use the transpose of this, the fst's will still run but your output will be garbage text.  

EG: 
 ```
model = torch_model.eval()
# outputs from softmax layer
outputs = model(inputs) 
# get log-likes
outputs_for_pykaldi = torch.log(outputs) 
 # detach from gpu if using one;  ix to get  shape = [frames, classes]
outputs_for_pykaldi = outputs_for_pykaldi.cpu().detatch().numpy()[ix, for, dims]
# call fn from above
words = decode_words_from_phones(outputs_for_pykaldi, *args) 
```
